crates/diem-client/src/state.rs

```diff
@@ -19,6 +21,35 @@ impl State {
             timestamp_usecs: resp.diem_ledger_timestampusec,
         }
     }
+
+    pub fn from_headers(headers: &reqwest::header::HeaderMap) -> anyhow::Result<Self> {
+        let maybe_chain_id = headers
+            .get(X_DIEM_CHAIN_ID)
+            .and_then(|h| h.to_str().ok())
+            .and_then(|s| s.parse().ok());
+        let maybe_version = headers
+            .get(X_DIEM_VERSION_ID)
+            .and_then(|h| h.to_str().ok())
+            .and_then(|s| s.parse().ok());
+        let maybe_timestamp = headers
+            .get(X_DIEM_TIMESTAMP_USEC_ID)
+            .and_then(|h| h.to_str().ok())
+            .and_then(|s| s.parse().ok());
+
+        let state = if let (Some(chain_id), Some(version), Some(timestamp_usecs)) =
+            (maybe_chain_id, maybe_version, maybe_timestamp)
+        {
+            Self {
+                chain_id,
+                version,
+                timestamp_usecs,
+            }
+        } else {
+            todo!()
```

Comment:
So, is this going to panic if you don't get the 3 headers?

We should follow up and do something here.